ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,906 @@
+                                    position: position.clone(),
+                                }),
+                            );
+                            let simulated_secret_proof: Hint = Hint::SecretProven(
+                                SecretProven::SimulatedSecretProof(SimulatedSecretProof {
+                                    image: leaf.proposition(),
+                                    challenge: leaf.challenge(),
+                                    unchecked_tree: UncheckedTree::UncheckedLeaf(leaf),
+                                    position,
+                                }),
+                            );
+                            bag.add_hint(simulated_commitment);
+                            bag.add_hint(simulated_secret_proof);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    match ut {
+        Ok(tree) => {
+            let mut bag: HintsBag = HintsBag::empty();
+
+            traverse_node(
+                tree,
+                real_propositions,
+                simulated_propositions,
+                NodePosition::crypto_tree_prefix(),
+                &mut bag,
+            );
+            Ok(bag)
+        }
+
+        Err(_) => Err("sigma parsing error"),
+    }
+}
+
+/// A method which is generating commitments to randomness. A commitment is about a first step
+/// of a zero-knowledge proof-of-knowledge knowledge protocol.
+/// the commitments are generated for every input box of transaction
+/// and return as `TransactionHintsBag`.
+/// generated commitments corresponds to `public_keys` that prepared in function input
+pub fn generate_commitments(
+    tx_context: TransactionContext<UnsignedTransaction>,
+    state_context: &ErgoStateContext,
+    public_keys: &[SigmaBoolean],
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let tx = tx_context.spending_tx.clone();
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = tx_context
+            .get_boxes_to_spend()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let ctx = Rc::new(make_context(state_context, &tx_context, i)?);
+        let tree = input_box.ergo_tree.clone();
+        let exp = tree
+            .proposition()
+            .map_err(ProverError::ErgoTreeError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+        let reduction_result = reduce_to_crypto(&exp, &Env::empty(), ctx)
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(i, generate_commitments_for(sigma_tree, public_keys));
+    }
+    Ok(hints_bag)
+}
+
+/// Extracting hints from a transaction and outputs it's corresponding TransactionHintsBag
+pub fn extract_hints(
+    tx: &Transaction,
```

Comment:
Let's use the updated `TransactionContext<Transaction>` instead of tx, boxes and data boxes.

Reply:
Yes, good point I rewrite that in the new commits!